Add projectRootDir setting

stefan-toubia opened the pull request:
Hello again.  This PR proposes a small but meaningful feature that will allow us users working in monorepos to configure our replConnectSequence with a specific project root directory.  Previously, I've always had to open a specific file before doing jack-in.  This addition removes that step, and a lot of headaches that came with it.

## What has Changed?
- Add projectRootDir setting to `calva.connectionSequences.menuSelections` to manually override the project root directory.  This is useful for monorepo projects.

Addressed #579

## My Calva PR Checklist
I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Made sure I am directing this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I am changed the default PR base branch, so that it is not `master`. (Sorry for the nagging.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.) You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. (For now you'll need to opt in to the CircleCI _New Experience_ UI to see the Artifacts tab, because bug.)
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.

## The Calva Team PR Checklist:
<!-- Please read the list, since you'll get a better idea about what to expect by doing so. 😄 -->

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse, @bpringe

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

stefan-toubia committed:
Add projectRootDir setting

PEZ commented:
Thanks! I'll have a look and test this. Can you also update the user docs for custom sequences? It is often easiest to judge if this kind of change is applied on the right place, by reading the user docs. It all looks good at my quick reading of the diffs, and I might have some more feedback in a moment.

stefan-toubia committed:
Update docs

stefan-toubia commented:
Updated.  My main area of uncertainty was exactly where to put this new property within the `replConnectionSequences`.  I'm not sure I understand the difference between what's inside `menuSelections` vs what's on the same level, but it seems to fit inside `menuSelections`.

PEZ commented:
`menuSelections` are meant for pre-filling/answering prompts that Calva would otherwise show the user. I think this fits better at the root level of the sequence, side by side with `name` and those.

The whole idea is a clever help for monorepos. I have been thinking quite a lot about how to support them better, but didn't think about how well it fits with a custom sequence. 😄

stefan-toubia committed:
Move rootProjectDir to root of replConnectSequences

stefan-toubia commented:
Sounds good to me, I've moved it to the root level.

I've been using this new feature for the past couple days and it's definitely been a small but nice quality of life improvement!  I can't speak for other monorepo users, but it works well for our code base.  We recently gone through the process of switching most of our clojure projects from lein to Clojure CLI, and implemented a custom `all/deps.edn` file that lists all dependencies in our code base.  Now doing jack-in on that directory lets us jump around all our projects without needing to restart.  We did, however, have to fork the Clojure CLI tool to make a small fix to support our monorepo workflow.

PEZ commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ Changes to Calva.
 ## [Unreleased]
 
 ## [2.0.80] - 2020-03-07
+- Add replConnectSequences.menuSelections.projectRootDir for manually specifying the root directory in order to support jack-in for monorepo projects.
```
Merge mistake? 😄 This should go under **Unreleased**. Also, I usually use the text and link for the issue I am fixing. So this would read:
```
- [Add option to manually specify project root directory](https://github.com/BetterThanTomorrow/calva/issues/579)
```

PEZ commented on package.json:
```diff
@@ -414,6 +414,10 @@
+                                        },
+                                        "projectRootDir": {
+                                            "type": "string",
+                                            "descripion": "Defines the project root directory.  Useful for working with monorepos."
```
I think it needs to be clear that there is a project root dir dynamically defined even without this setting. As it is worded now, for someone just reading the first sentence (yes, those people exist 😄 ) it can leave the impression that this setting is needed. Consider pasting the text you wrote for the user docs. It makes the purpose of the setting crystal clear, imo.

stefan-toubia replied:
Good call, fixed.

PEZ commented on src/nrepl/jack-in.ts:
```diff
@@ -41,6 +41,10 @@ function cancelJackInTask() {
 async function executeJackInTask(projectType: projectTypes.ProjectType, projectTypeSelection: any, executable: string, args: any, cljTypes: string[], outputChannel: vscode.OutputChannel, connectSequence: ReplConnectSequence) {
+    // If the connection sequence specifies a project root dir, override the existing project dir
+    const projectRootDir = connectSequence?.menuSelections?.projectRootDir;
+    if(projectRootDir) state.setProjectRoot(projectRootDir);
```
Minor detail. The code base might not show it, but I am trying to move in the direction where statement bodies of `if`s and such are enclosed in curly brackets. See https://github.com/BetterThanTomorrow/calva/wiki/Coding-Style#curlies

stefan-toubia replied:
Works for me, should have read the style guide more thoroughly.

PEZ commented on src/state.ts:
```diff
@@ -156,6 +156,10 @@ export function getProjectRoot(useCache = true): string {
     }
 }
 
+export function setProjectRoot(rootPath: string) {
```
👍

stefan-toubia committed:
PR fixes

PEZ reviewed (approved):
This is correct. 😄

PEZ commented:
So, now I have merged things to dev that also deals with the project dir thing. Can you pull that in and see that your changes still work? Sorry about the hassle!

stefan-toubia commented:
Sorry for the delay, things have been a bit hectic with all this virus business.  I will check this soon.

PEZ commented:
No problem at all!

Here's a doc page that might benefit from a mention of/section about monorepos: https://calva.readthedocs.io/en/latest/workspace-layouts.html

mrkam2 commented:
@stefan-toubia bump :-)

PEZ commented:
I think things have moved quite a bit around this in Calva. Not sure which route is best, but maybe it is easiest to start over from a fresh `dev` and apply the relevant changes manually.

stefan-toubia commented:
Closing this, replaced by #1100
